pdftext/pdf/utils.py

```diff
@@ -1,98 +1,36 @@
+from ctypes import byref, c_int, create_string_buffer
+
+import pypdfium2 as pdfium
 import pypdfium2.raw as pdfium_c
-import ctypes
 
 LINE_BREAKS = ["\n", "\u000D", "\u000A"]
 TABS = ["\t", "\u0009", "\x09"]
 SPACES = [" ", "\ufffe", "\uFEFF", "\xa0"]
 WHITESPACE_CHARS = ["\n", "\r", "\f", "\t", " "]
 
 
-def unnormalize_bbox(bbox, page_width, page_height):
-    x1 = round(bbox[0] * page_width, 1)
-    y1 = round(bbox[1] * page_height, 1)
-    x2 = round(bbox[2] * page_width, 1)
-    y2 = round(bbox[3] * page_height, 1)
-    return x1, y1, x2, y2
+def flatten(page, flag=pdfium_c.FLAT_NORMALDISPLAY):
+    rc = pdfium_c.FPDFPage_Flatten(page, flag)
+    if rc == pdfium_c.FLATTEN_FAIL:
+        raise pdfium.PdfiumError("Failed to flatten annotations / form fields.")
 
 
-def get_fontname(textpage, char_index):
-    n_bytes = pdfium_c.FPDFText_GetFontInfo(textpage, char_index, None, 0, None)
-    buffer = ctypes.create_string_buffer(n_bytes)
-    # Re-interpret the type from char to unsigned short as required by the function
-    buffer_ptr = ctypes.cast(buffer, ctypes.POINTER(ctypes.c_ushort))
-    flag_buffer = ctypes.c_int()
-    font_info = pdfium_c.FPDFText_GetFontInfo(textpage, char_index, buffer_ptr, n_bytes, flag_buffer)
-    if font_info == 0:
-        return None, None
+def get_fontname(textpage, i):
```

Comment:
Looks like you're getting this for every char - I was sampling before due to the performance hit.  Is this not a big deal now since it's faster overall?

Reply:
The previous implementation first obtained the buffer size and then retrieved the font name, resulting in two calls for any character. The current implementation uses a 256-byte buffer and retrieves the font name in a single call, falling back only if the created buffer is too small (which I haven't observed so far). We also rely on this to break spans now.